capsules/src/hmac.rs

```diff
@@ -0,0 +1,488 @@
+//! HMAC (Hash-based Message Authentication Code).
```

Comment:
This is a tricky `Driver` to write with respect to virtualization. There is a key decision to make:
- Apps have to allow the entire buffer to be HMACd.
or
- Apps can can call "add data" like the HIL allows, and only allow a much smaller buffer than the actual data to be HMACd.

If the latter, we have to allow multiple syscall/callback cycles, yet since the data have to go to the HW, the app effectively "owns" the HMAC HW during those loops. What happens if an app indicates it wants to do a really long HMAC, but then stops providing data? The `Driver` needs some way to reclaim the hardware. I don't think we have another `Driver` currently that presents this issue.

I think for now, using the former option and setting up the driver like our other capsules (ninedof for example) is a reasonable compromise. That would mean restructuring this capsule slightly so that there is an "enqueue_command" function that is triggered on a run command that iteratively executes HMACs it if multiple apps request them.

Reply:
The idea is that whatever data buffer the app passes to the driver will be HMACd. I don't anticipate a use case where an app only wants to HMAC part of the data buffer. We currently only allow a single data buffer to be used, although that might be changeable in the future.

What I am picturing (untested) is that two apps both create the buffers they want and call the `allow` syscalls to tell the driver. One app will call the `run` syscall first and start the HMAC process. While this is happening if the other app calls `run` they will get an error and have to retry. Once the first app has finished then the second app can call `run` and work successfully.

You think the `run` logic should be changed to allow a queue of run commands so we don't return an error. I'll give that a go.

Reply:
> You think the `run` logic should be changed to allow a queue of run commands so we don't return an error. I'll give that a go.

I think this is the correct approach. Alternatively, to make the apps aware of the possible delay, we could keep the current approach and introduce a subscription on the event that the driver is available again. That would remove the need for complex queuing and give the apps more control.